examples/wgpu.rs

```diff
@@ -0,0 +1,227 @@
+        depth_stencil: None,
+        multisample: wgpu::MultisampleState::default(),
+        multiview: None,
+    });
+
+    let config = wgpu::SurfaceConfiguration {
+        usage: wgpu::TextureUsages::RENDER_ATTACHMENT,
+        format: swapchain_format,
+        width: size.width,
+        height: size.height,
+        present_mode: wgpu::PresentMode::Fifo,
+        desired_maximum_frame_latency: 2,
+        alpha_mode: wgpu::CompositeAlphaMode::PostMultiplied,
+        view_formats: vec![],
+    };
+
+    surface.configure(&device, &config);
+    (surface, device, render_pipeline, queue)
+}
+
+fn render_triangle(
+    surface: &wgpu::Surface,
+    device: &wgpu::Device,
+    render_pipeline: &wgpu::RenderPipeline,
+    queue: &wgpu::Queue,
+) {
+    let frame = surface
+        .get_current_texture()
+        .expect("Failed to acquire next swap chain texture");
+    let view = frame
+        .texture
+        .create_view(&wgpu::TextureViewDescriptor::default());
+    let mut encoder =
+        device.create_command_encoder(&wgpu::CommandEncoderDescriptor { label: None });
+    {
+        let mut rpass = encoder.begin_render_pass(&wgpu::RenderPassDescriptor {
+            label: None,
+            color_attachments: &[Some(wgpu::RenderPassColorAttachment {
+                view: &view,
+                resolve_target: None,
+                ops: wgpu::Operations {
+                    load: wgpu::LoadOp::Clear(wgpu::Color::TRANSPARENT),
+                    store: wgpu::StoreOp::Store,
+                },
+            })],
+            depth_stencil_attachment: None,
+            timestamp_writes: None,
+            occlusion_query_set: None,
+        });
+        rpass.set_pipeline(&render_pipeline);
+        rpass.draw(0..3, 0..1);
+    }
+
+    queue.submit(Some(encoder.finish()));
+    frame.present();
+}
+
+struct GraphicsResources<'a> {
+    surface: wgpu::Surface<'a>,
+    device: wgpu::Device,
+    pipeline: wgpu::RenderPipeline,
+    queue: wgpu::Queue,
+}
+
+#[self_referencing]
+struct GraphicsContext {
+    desktop: DesktopContext,
+    #[borrows(desktop)]
+    #[not_covariant]
+    resources: GraphicsResources<'this>,
+}
+
+fn app() -> Element {
+    let mut graphics_resources = use_resource(move || async {
+        GraphicsContextAsyncBuilder {
+            desktop: window(),
+            resources_builder: |desktop: &DesktopContext| {
+                Box::pin(async move {
+                    let (surface, device, pipeline, queue) = setup_triangle(desktop).await;
+                    GraphicsResources {
+                        surface: surface,
+                        device: device,
+                        pipeline: pipeline,
+                        queue: queue,
+                    }
+                })
+            },
+        }
+        .build()
+        .await
+    });
+
+    let desktop_context = use_window();
+    graphics_resources.read().as_ref().map(|_| {
+        desktop_context.window.request_redraw();
```

Comment:
Should we be doing this on every render? Maybe a use_effect or whenever the contents change (in say, the wry_event_handler)